frontend/src/hocs/withScreencast/withScreencast.tsx

```diff
@@ -0,0 +1,267 @@
+  fileName: string
+  recording: boolean
+  recordAudio: boolean
+  countdown: number
+  startAnimation: boolean
+  blob?: Blob
+  showRecordedDialog: boolean
+  showScreencastDialog: boolean
+  showUnsupportedDialog: boolean
+}
+
+export interface ScreenCastHOC extends WithScreenCastState {
+  toggleRecordAudio: () => void
+  startRecording: (fileName: string) => void
+  stopRecording: () => void
+}
+
+const SUPPORTED_BROWSERS: any = {
+  Chrome: "49",
+  Firefox: "29",
+  Opera: "36",
+  Edge: "76",
+}
+
+function withScreencast(
+  WrappedComponent: ComponentType<any>
+): ComponentType<any> {
+  class ComponentWithScreencast extends PureComponent<
+    WithScreenCastProps,
+    WithScreenCastState
+  > {
+    static readonly displayName = `withScreencast(${WrappedComponent.displayName ||
+      WrappedComponent.name})`
+
+    recorder?: ScreenCastRecorder | null
+
+    state = {
+      fileName: "streamlit-screencast",
+      countdown: -1,
+      recording: false,
+      recordAudio: false,
+      startAnimation: false,
+      showRecordedDialog: false,
+      showScreencastDialog: false,
+      showUnsupportedDialog: false,
+    }
+
+    toggleRecordAudio = (): void => {
+      const { recordAudio } = this.state
+
+      this.setState({
+        recordAudio: !recordAudio,
+      })
+    }
+
+    showDialog = (fileName: string): void => {
+      const { recording, countdown } = this.state
+
+      if (!this.checkSupportedBrowser()) {
+        this.setState({
+          showUnsupportedDialog: true,
+        })
+      } else {
+        if (!recording && countdown < 0) {
+          this.setState({
+            fileName,
+            showScreencastDialog: true,
+          })
+        } else {
+          this.stopRecording()
+        }
+      }
+    }
+
+    startRecording = async (): Promise<any> => {
+      const { recordAudio } = this.state
+      this.recorder = new ScreenCastRecorder({ recordAudio })
+
+      await this.recorder.initialize()
+
+      this.setState({
+        countdown: 3,
+        startAnimation: true,
+        showRecordedDialog: false,
+      })
+    }
+
+    stopRecording = async (): Promise<any> => {
+      let blob
+      const { recording } = this.state
+
+      if (
+        this.recorder &&
+        recording &&
+        this.recorder.getState() !== "inactive"
+      )
+        blob = await this.recorder.stop()
+
+      this.setState({
+        blob,
+        countdown: -1,
+        recording: false,
+        startAnimation: false,
+        showRecordedDialog: true,
+      })
+    }
+
+    onAnimationEnd = async (): Promise<any> => {
+      const { countdown } = this.state
+
+      this.setState({
+        startAnimation: false,
+        countdown: countdown - 1,
+      })
+
+      if (countdown - 1 > 0) {
+        window.setTimeout(() => {
+          this.setState({
+            startAnimation: true,
+          })
+        }, 1000)
+      }
+
+      if (countdown - 1 === 0 && this.recorder) {
```

Comment:
Why would`this.recorder` be `null`?

Or did you add this check just for Typescript's sake? If the latter, then it's better to do something like this instead:

```
if (this.recorder == null) {
  // Should never happen.
  throw new Error("Countdown finished but recorder is null")
}
```